include/aspect/material_model/averaging.h

```diff
@@ -0,0 +1,148 @@
+/*
+  Copyright (C) 2014 by the authors of the ASPECT code.
+  This file is part of ASPECT.
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#ifndef __aspect__averaging_h
+#define __aspect__averaging_h
+
+#include <deal.II/base/function_lib.h>
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <deal.II/base/parsed_function.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+
+
+    /**
+     * A material model that applies a depth-dependent viscosity to a ''base model''
+     * chosen from any of the other available material models. This depth-dependent
+     * material model allows the user to specify a depth-dependent reference viscosity
+     * either through a parsed function, lists of depth and viscosity values, or a file.
+     * The current implementation only allows for depth-dependence of viscosity - all
+     * other properties are derived from the base model.
+     * @ingroup MaterialModels
+     */
+
+    template <int dim>
+    class Averaging : public MaterialModel::Interface<dim>, public ::aspect::SimulatorAccess<dim>
+    {
+      public:
+        /**
+                 * Given the averaging @p operation, a description of where the
+                 * quadrature points are located on the given cell, and a mapping,
+                 * perform this operation on all elements of the @p values structure.
+                 */
+        virtual
+        void
+        average (const std::string &averaging_operation,
+                 const std::vector<Point<dim> >    &position,
+                 std::vector<double>           &values_out) const;
```

Comment:
Does this function have to virtual?

It is also essentially an internal function, so you should be able to put it into the private section of the class.